Common/src/main/java/com/techarium/techarium/multiblock/MultiBlockStructure.java

```diff
@@ -0,0 +1,265 @@
+package com.techarium.techarium.multiblock;
+
+import com.google.common.base.Joiner;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import com.techarium.techarium.block.multiblock.MachineCoreBlock;
+import com.techarium.techarium.block.selfdeploying.SelfDeployingBlock;
+import com.techarium.techarium.blockentity.selfdeploying.SelfDeployingMultiBlockBlockEntity;
+import com.techarium.techarium.platform.CommonServices;
+import com.techarium.techarium.registry.TechariumBlocks;
+import com.techarium.techarium.util.BlockRegion;
+import com.techarium.techarium.util.MathUtils;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.properties.BlockStateProperties;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Define a multiblock structure.
+ * Positions of the elements are considered offset from the core position, with the core oriented toward south, and in the minecraft axis (south:+z, east:+x).
+ * <br>
+ * Instances of this class are created via datapack.
+ */
+public class MultiBlockStructure {
```

Comment:
It would be cleaner and not need a manual `toString` implementation.

Reply:
the problem I have with records is that they are final class, so if someone want to extend this class (I don't have examples) htey won't be able to.

Reply:
Well, you would want to have an interface that the class implements such as `MultiblockStructure`, and then name this something like `SimpleMultiblockStructure` because if your plan was to allow for people to be able to extend and override the methods there is a lot of storage overhead in that class that cannot be gotten rid of due to the default implementation.